### ISSUE TITLE
something not right with the U objective function in variational.py

### ISSUE THREAD

**Issue by bhomass** [wohlert/semi-supervised-pytorch#28]
it seems to me line 105 in variational.py is a mistake

L = L.view_as(logits.t()).t()

The L according to eqn 7 in "Semi-supervised Learning with Deep Generative Models" is meant to be the expectation value of the Labeled objective function, which is the entire eqn 6 in that paper. Here you redefined L to be simply derived from the logit output of model.classify(). 

Please verify you have the correct formula. If so, where is the paper reference for it?

**Comment by wohlert** [comment]
There is no mistake. The function `L = L.view_as(example)` returns exactly `L`, but reshaped to have the same dimensions as `example`. In the following line [109](https://github.com/wohlert/semi-supervised-pytorch/blob/master/semi-supervised/inference/variational.py#L109) we multiply L by the logits as described in the paper.

![](https://latex.codecogs.com/svg.latex?%5Csum_yq_%7B%5Ctheta%7D%28y%7Cx%29%28-L%28x%2Cy%29%29)

Note that I have tried to very hard to make the code particularly understandable by following the equations in the papers closely, and I believe this library provides that to a much higher degree than other available implementations. However due to limitations of software, we cannot always map equations 1-to-1 to code.

If you have further questions regarding this, then maybe this issue https://github.com/wohlert/semi-supervised-pytorch/issues/15 will help you.

In the future please refrain from creating issues that are not directly related to the code before having consulted either existing issues, mathoverflow, papers or the likes.

**Comment by bhomass** [comment]
yes, on second read, I see you are using L from the labeled part. Sorry to post in haste. Yes, the code is very easy to match to formula. I commend you on that.
